crates/wassette/src/lib.rs

```diff
@@ -198,577 +300,241 @@ pub struct ComponentInstance {
+    ) -> Result<PathBuf> {
+        let target_path = self.component_path(component_id);
+        match resource {
+            DownloadedResource::Local(path) if path == target_path => Ok(target_path),
+            other => {
+                let _permit = self.storage.acquire_download_permit().await;
+                // Remove any stale component to avoid partial states when copying
+                self.storage
+                    .remove_if_exists(&target_path, "component file", component_id)
+                    .await?;
+                self.storage
+                    .remove_if_exists(
+                        &self.storage.metadata_path(component_id),
+                        "component metadata file",
+                        component_id,
+                    )
+                    .await?;
+                self.storage
+                    .remove_if_exists(
+                        &self.storage.precompiled_path(component_id),
+                        "precompiled component file",
+                        component_id,
+                    )
+                    .await?;
+                other.copy_to(self.storage.root()).await.with_context(|| {
+                    format!(
+                        "Failed to copy component to destination: {}",
+                        self.storage.root().display()
+                    )
+                })?;
+                Ok(target_path)
+            }
         }
+    }
 
-        if let Err(e) = downloaded_resource.copy_to(&self.plugin_dir).await {
-            let mut registry_write = self.registry.write().await;
-            registry_write.unregister_component(&id);
-            bail!(
-                "Failed to copy component to destination: {}. Error: {}",
-                self.plugin_dir.display(),
-                e
-            );
-        }
+    async fn compile_and_register_component(
+        &self,
+        component_id: &str,
+        wasm_path: &Path,
+    ) -> Result<ComponentLoadOutcome> {
+        let (component, _wasm_bytes) = self
+            .load_component_optimized(wasm_path, component_id)
+            .await?;
+
+        let instance_pre = self
+            .runtime
+            .instantiate_pre(&component)
+            .context("failed to instantiate component")?;
+
+        let component_instance = ComponentInstance {
+            component: Arc::new(component),
+            instance_pre: Arc::new(instance_pre),
+        };
+
+        let tool_metadata =
+            component_exports_to_tools(&component_instance.component, self.runtime.engine(), true);
+        let tool_names: Vec<String> = tool_metadata
+            .iter()
+            .map(|tool| tool.normalized_name.clone())
+            .collect();
 
-        // Save metadata for future startups
         if let Ok(validation_stamp) =
-            Self::create_validation_stamp(&self.component_path(&id), false).await
+            ComponentStorage::create_validation_stamp(wasm_path, false).await
```

Comment:
Nit: It is a bit weird that this is a static method now that I'm looking at it. Maybe just have this be on `self` instead?